common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SwitchBrowserUriHelper.kt

```diff
@@ -0,0 +1,118 @@
+     * @return The process uri constructed from the broker redirect uri.
+     * e.g. your-action-uri?code=your-switch-browser-code
+     */
+    fun buildProcessUri(brokerRedirectUri: Uri): Uri? {
+        val methodTag = "$TAG:buildProcessUri"
+        // Get the session token from the broker redirect uri.
+        val sessionToken = brokerRedirectUri.getQueryParameter(
+            SWITCH_BROWSER.CODE
+        )
+        if (sessionToken.isNullOrEmpty()) {
+            // This should never happen, but if it does, we should log it and return.
+            Logger.warn(methodTag, "Switch browser code is null or empty ")
+            return null
+        }
+        // Get the process uri from the broker redirect uri.
+        val actionUri = brokerRedirectUri.getQueryParameter(
+            SWITCH_BROWSER.ACTION_URI
+        )
+        if (actionUri.isNullOrEmpty()) {
+            // This should never happen, but if it does, we should log it and return.
+            Logger.warn(methodTag, "Switch browser action URI is null or empty ")
+            return null
+        }
+        // Query parameters for the process uri.
+        val queryParams = hashMapOf<String, String>()
+        queryParams[SWITCH_BROWSER.CODE] = sessionToken
+        queryParams[OAuth2.REDIRECT_URI] = Broker.NEW_BROKER_REDIRECT_URI
+        // Construct the uri to the process endpoint.
+        return buildSwitchBrowserUri(actionUri, queryParams)
+    }
+
+    /**
+     * Build a generic switch browser uri.
+     *
+     * @param actionUri The action uri to be opened.
+     * @param queryParams The query parameters to be included in the switch browser uri.
+     *
+     * @return The switch browser uri constructed from the action uri and query parameters.
+     */
+    private fun buildSwitchBrowserUri(
+        actionUri: String,
+        queryParams: HashMap<String, String>
+    ): Uri? {
+        val paths = actionUri.split("/")
+        val authority = paths[0]
+        val uriBuilder = Uri.Builder()
+            .scheme("https")
+            .encodedAuthority(authority)
+        for (i in 1 until paths.size) {
+            uriBuilder.appendPath(paths[i])
+        }
+        for ((key, value) in queryParams.entries) {
+            uriBuilder.appendQueryParameter(key, value)
+        }
+        return uriBuilder.build()
+    }
+
+    /**
+     * Check if the request is to switch the browser.
+     *
+     * The request is considered "switch_browser" if the URL contains
+     * the action URI, code, and action parameters.
+     *
+     * @param uri The URI of the request.
+     * @return True if the request contains the required parameters, false otherwise.
+     */
+    fun isSwitchBrowserRequest(uri: Uri?): Boolean {
+        if (uri == null) {
+            return false
+        }
+        val requiredParams =
+            setOf(SWITCH_BROWSER.ACTION_URI, SWITCH_BROWSER.CODE, SWITCH_BROWSER.ACTION)
+        return uri.queryParameterNames.containsAll(requiredParams)
```

Comment:
The redirect path has slightly changed on the server side and that should make it easier to detect switch browser flow for clients. Previously we redirected {redirectUri}?action=[action]&action_uri=[action_uri]&code=[code]

but now we redirect to following:

{redirectUri}/switch_browser?action_uri=[action_uri]&code=[code]

Note the addition of the path `/switch_browser` after redirect uri and omission of the `action` qp. The `/switch_browser` path itself should be enough to detect this is a switch browser flow. 

Once detected that this is switch_browser flow, then it likely still makes sense to validate the presence of action_uri and code but then the absence of those should result in an exception to be thrown because we would be certain it is switch browser flow.

Reply:
logic updated